architecture-model-generator/modules/architecture-model-generator-core/src/main/java/io/ballerina/artifactsgenerator/ModuleNodeTransformer.java

```diff
@@ -0,0 +1,246 @@
+ * transforming nodes from a different module, as the semantic model context is specific to the module being processed.
+ * </p>
+ *
+ * @since 2.3.0
+ */
+public class ModuleNodeTransformer extends NodeTransformer<Optional<Artifact>> {
+
+    private final SemanticModel semanticModel;
+
+    private static final String AUTOMATION_FUNCTION_NAME = "automation";
+    private static final String MAIN_FUNCTION_NAME = "main";
+    private static final String BALLERINAX_ORG_NAME = "ballerinax";
+    private static final String NP_MODULE_NAME = "np";
+    private static final String NP_FUNCTION_ANNOTATION = "NaturalFunction";
+
+    public ModuleNodeTransformer(SemanticModel semanticModel) {
+        this.semanticModel = semanticModel;
+    }
+
+    @Override
+    public Optional<Artifact> transform(FunctionDefinitionNode functionDefinitionNode) {
+        Artifact.Builder functionBuilder = new Artifact.Builder(functionDefinitionNode);
+        String functionName = functionDefinitionNode.functionName().text();
+
+        if (functionName.equals(MAIN_FUNCTION_NAME)) {
+            functionBuilder
+                    .name(AUTOMATION_FUNCTION_NAME)
+                    .type(Artifact.Type.AUTOMATION);
+        } else if (functionDefinitionNode.functionBody().kind() == SyntaxKind.EXPRESSION_FUNCTION_BODY) {
+            functionBuilder
+                    .name(functionName)
+                    .type(Artifact.Type.DATA_MAPPER);
+        } else if (functionDefinitionNode.kind() == SyntaxKind.RESOURCE_ACCESSOR_DEFINITION) {
+            functionBuilder
+                    .accessor(functionName)
+                    .name(getPathString(functionDefinitionNode.relativeResourcePath()))
+                    .type(Artifact.Type.RESOURCE);
+        } else if (hasQualifier(functionDefinitionNode.qualifierList(), SyntaxKind.REMOTE_KEYWORD)) {
+            functionBuilder
+                    .name(functionName)
+                    .type(Artifact.Type.REMOTE);
+        } else if (isPromptAsCodeFunction(functionDefinitionNode)) {
+            functionBuilder
+                    .name(functionName)
+                    .type(Artifact.Type.NP_FUNCTION);
+        } else {
+            functionBuilder
+                    .name(functionName)
+                    .type(Artifact.Type.FUNCTION);
+        }
+        return Optional.of(functionBuilder.build());
+    }
+
+    @Override
+    public Optional<Artifact> transform(ServiceDeclarationNode serviceDeclarationNode) {
+        Artifact.Builder serviceBuilder = new Artifact.Builder(serviceDeclarationNode).locationId();
+
+        // Set the icon using the listener
+        SeparatedNodeList<ExpressionNode> expressions = serviceDeclarationNode.expressions();
+        if (!expressions.isEmpty()) {
+            setIcon(serviceBuilder, expressions.get(0));
+        }
+
+        // Generate the service path
+        String absolutePath = getPathString(serviceDeclarationNode.absoluteResourcePath());
```

Comment:
Here only for `graphql`, `http` and `ai:Service` has a abosoluteResourcePath. For RabbitMQ we can have a strign-literal. For other we should resolve service type using the attached ServiceType or via the Listener